project-koku/korekuta-operator-go
# Add upload variables for toggling upload/upload cycle

## Description 
This PR adds the following to the CR spec: 
* UploadCycle - the schedule we are uploading on in minutes - defaults to 360 min
* UploadToggle - a boolean stating whether or not we are uploading - defaults to true

This PR adds the following to the CR status: 
* OperatorCommit - the commit hash of the operator 
* UploadCycle - the schedule we are uploading on in minutes
* UploadToggle - a boolean stating whether or not we are uploading 
* LastUploadStatus - a string with the response code and status
* LastUploadTime - the last attempted upload time
* LastSuccessfulUploadTime - the last successful upload time 

It also does the following: 
* changes to using the ca-bundle.crt for uploading to cloud.redhat.com 
* Adds logic to check the UploadToggle spec field to see if a customer has specified not to upload 
* Adds logic to check and see when the LastSuccessfulUploadTime was, and check whether or not it has been longer/shorter than the specified UploadCycle. 

## Testing:
1. Locally: 
* Deploy the CR to the cluster 
* create the /tmp/cost-mgmt-operator-reports/ directory & copy a tarfile to that location.
* you will need to copy the ca-bundle that the operator creates locally - you can reach out to me for the file if you don't feel like deploying the operator to the cluster first to get the bundle. This ca-bundle.crt file needs to exist at `/var/run/configmaps/trusted-ca-bundle/ca-bundle.crt`

2. Cluster: 
* Deploy the operator and the CR to the cluster
* Watch the operator logs - reach out to me for a tar.gz (or you can use any tar.gz to upload).
If you are running in the cluster you can copy the tarfile to the pod using this command (You will replace the podname with your specific podname):
kubectl cp payload1.tar.gz korekuta-operator-go-controller-manager-5ffd598cbd-678fp:/tmp/cost-mgmt-operator-reports/ -c manager
* Watch the operator logs

## Testing Scenarios: 
* Set the upload_toggle in the CR spec to false - make sure that upload doesn't run (you should see a log that the operator is not configured to upload) 
* Set the upload_toggle to true - make sure that upload attempts to run 
* change the upload_cycle to X min and make sure that once X min has passed since the last successful upload, that upload is ran again. 


## Note 
* I will add unit tests after I merge this PR and get the split payload logic in place so that I can add unit tests for the entire issue functionality
